C7Engine/MapUnitExtensions.cs

```diff
@@ -102,53 +97,111 @@ public static void RollToPromote(this MapUnit unit, MapUnit opponent, bool waitF
+	}
+
+	public static CombatResult fight(this MapUnit attacker, MapUnit defender)
 	{
 		// Rotate defender to face its attacker. We'll restore the original facing direction at the end of the battle.
 		var defenderOriginalDirection = defender.facingDirection;
-		defender.facingDirection = unit.facingDirection.reversed();
+		defender.facingDirection = attacker.facingDirection.reversed();
 
-		IEnumerable<StrengthBonus> attackBonuses  = unit    .ListStrengthBonusesVersus(defender, true , false, unit.facingDirection),
-		                           defenseBonuses = defender.ListStrengthBonusesVersus(unit    , false, false, unit.facingDirection);
+		IEnumerable<StrengthBonus> attackBonuses  = attacker.ListStrengthBonusesVersus(defender, CombatRole.Attack , attacker.facingDirection),
+		                           defenseBonuses = defender.ListStrengthBonusesVersus(attacker, CombatRole.Defense, attacker.facingDirection);
 
-		double attackerStrength = unit    .unitType.attack  * StrengthBonus.ListToMultiplier(attackBonuses),
+		double attackerStrength = attacker.unitType.attack  * StrengthBonus.ListToMultiplier(attackBonuses),
 		       defenderStrength = defender.unitType.defense * StrengthBonus.ListToMultiplier(defenseBonuses);
 
-		Console.WriteLine($"Combat log: {unit.unitType.name} ({attackerStrength}) attacking {defender.unitType.name} ({defenderStrength})");
-		Console.WriteLine($"\tAttacker: {unit.unitType.name}, base strength {unit.unitType.attack}");
+		Console.WriteLine($"Combat log: {attacker.unitType.name} ({attackerStrength}) attacking {defender.unitType.name} ({defenderStrength})");
+		Console.WriteLine($"\tAttacker: {attacker.unitType.name}, base strength {attacker.unitType.BaseStrength(CombatRole.Attack)}");
 		foreach (StrengthBonus bonus in attackBonuses)
 			Console.WriteLine($"\t\t+{100.0*bonus.amount}%\t{bonus.description}");
-		Console.WriteLine($"\tDefender: {defender.unitType.name}, base strength {defender.unitType.defense}");
+		Console.WriteLine($"\tDefender: {defender.unitType.name}, base strength {defender.unitType.BaseStrength(CombatRole.Defense)}");
 		foreach (StrengthBonus bonus in defenseBonuses)
 			Console.WriteLine($"\t\t+{100.0*bonus.amount}%\t{bonus.description}");
 
+		CombatResult result = CombatResult.Impossible;
+
 		double attackerOdds = attackerStrength / (attackerStrength + defenderStrength);
 		if (Double.IsNaN(attackerOdds))
-			return false;
+			return result;
+
+		// Defensive bombard
+		MapUnit defensiveBombarder = MapUnit.NONE;
+		double defensiveBombarderStrength = 0.0;
+		foreach (MapUnit candidate in defender.location.unitsOnTile.Where(u => u != defender && ! u.hasExhaustedDefensiveBombard)) {
+			double strength = candidate.StrengthVersus(attacker, CombatRole.DefensiveBombard, attacker.facingDirection.reversed());
+			if (strength > defensiveBombarderStrength) {
+				defensiveBombarder = candidate;
+				defensiveBombarderStrength = strength;
+			}
+		}
```

Comment:
eventually, it's probably worth converting all these search & compare cases to something more LINQ based (map/reduce, sort by, take first, and whatever else) and put them in utility functions